### DISCUSSION TITLE
文件->ilogtail->kafka 数据丢失

### DISCUSSION THREAD

**Discussion by q121102302** [alibaba/ilogtail#867] [Help] [answered]
版本：1.3.1
配置：
[file_kafka.txt](https://github.com/alibaba/ilogtail/files/11531333/file_kafka.txt)
[ilogtail_config.txt](https://github.com/alibaba/ilogtail/files/11531335/ilogtail_config.txt)
ilogtail日志：
![lQLPJwBeb5zzP27NAznNBRWwsn7YYRySdowEYnamugASAA_1301_825](https://github.com/alibaba/ilogtail/assets/101638329/4a91ea44-8162-43c3-91e8-647af7a983ae)
![lQLPJxfIAY4FH27NAeXNBfWw9zbxlBz2MYQEYnamvAATAA_1525_485](https://github.com/alibaba/ilogtail/assets/101638329/484d2b90-48a9-43ad-b5c4-571129f0543f)
问题：
使用过程中发现 有机器a丢失其中三秒数据 但是ilogtail并未有相应日志  
排查过程看了kafka的对应topic数据 未发现kafka中有这三条数据  
说明数据应该是在ilogtail端前后丢失 
这种问题应该怎么排查？
数据量的话 大约每秒100条的样子  日志文件大小为60m 大约15分钟更新一个新文件 即 4mb/min

**Comment by linrunqi08** [comment]
1. 完整的日志文件麻烦提供一下，ilogtail.LOG, logtail_plugin.LOG
2. 丢失的是哪三条数据，原始日志可以提供一下吗？具体的时间点是多少

**Reply by q121102302** [reply]
ilogtail.LOG, logtail_plugin.LOG日志文件：
[logtail_plugin.LOG](https://github.com/alibaba/ilogtail/files/11564343/logtail_plugin.LOG)
[ilogtail.LOG](https://github.com/alibaba/ilogtail/files/11564345/ilogtail.LOG)
原始日志文件：
[app.log](https://github.com/alibaba/ilogtail/files/11564346/app.log)
上文中的ilogtail_config.json配置貌似贴的有问题 我们配置的是：
`{
    "max_read_buffer_size": 4194304
}`

问题：丢失的是2023-05-22 16:31:05.271至2023-05-22 16:31:08.303的那一段大约3秒钟的日志  
已查过输出的topic 里面确实没有这台机器的这一段时间的日志
查看ilogtail进程 也并未有重启

**Reply by q121102302** [reply]
当时怀疑是 cpu或者内存过限了 然后ilogtail_config.json修改为：
`{
    "cpu_usage_limit": 1,
    "mem_usage_limit": 1024,
    "max_read_buffer_size": 4194304
}`
上传同样的文件 输出这条大日志时 依然还是会丢失

**Reply by linrunqi08** [reply]
@q121102302  看了下2023-05-22 16:31:05.227 这一条数据有8M多，现在是只有这样的大数据会丢失吗

**Reply by q121102302** [reply]
目前发现的有： 
1.一个文件中含有一条大日志的话  这个文件的这条大日志后的几秒钟的日志会跟着全部丢失。
2.在另一个测试环境中，ilogtail会读取多个日志文件，也发现过其中一个日志文件丢失了2～3秒中的数据，这个日志文件中并未发现大日志。但是没有对这台机器的其他日志文件进行排查，所以不确定是否是这台机器的其他日志文件中含有大数据影响了它。所以并未展示。

**Reply by linrunqi08** [reply]
系统配置里可以设置下这个参数试一下，max_read_buffer_size，单位是B，比如512 KB，就设置成max_read_buffer_size: 524288。你这个可以设置成10MB看下

**Reply by q121102302** [reply]
这个参数 我看logtail中介绍 最大是8388608  我们那个测试环境 用的就是8388608这个值。。
flushers中的MaxMessageBytes也设置成了439430400 依然遇到了情况2 所以感觉不像是这块儿的问题

**Reply by linrunqi08** [reply]
1. 8388608是商业版本的限制，开源这块可以设置更大的。调大之后，情况1是否就不存在了？
2. 情况2的日志以及配置等可以再提供一下吗

**Reply by q121102302** [reply]
1.我去调整到10m 然后观察下
2.日志当时未保留。。所以应该已经不在了 如果下次出现 再进行反馈吧

**Reply by q121102302** [reply]
我调大max_read_buffer_size 这个参数 那么flushers中的kafka_v2下的MaxMessageBytes这个参数是不是也要相应跟着调大

**Reply by q121102302** [reply]
情况一 调整为10m后 依然会丢失数据 出现的情况是：刚重启时 尝试上传了一个8.45Mb的文件 显示是成功的；今天再使用时，依然是失败的。
日志文件中 红框内是丢失的日志 而第一条 就是那个大日志文件 该条数据大约5.5Mb
<img width="1532" alt="image" src="https://github.com/alibaba/ilogtail/assets/101638329/f39d8b9f-9cc2-4ce2-acd9-1c99f7d1ab4c">

可以看到es日志直接从2023-05-29 17:13:39.478跳到了2023-05-29 17:13:43.212
<img width="874" alt="image" src="https://github.com/alibaba/ilogtail/assets/101638329/c0b74ffc-82ff-4958-8288-7c2e02b87649">


日志：
logtail_plugin.LOG
<img width="1632" alt="image" src="https://github.com/alibaba/ilogtail/assets/101638329/3a79bd74-ba30-406a-9c72-6ac88f2f42f4">
ilogtail.LOG
<img width="1621" alt="image" src="https://github.com/alibaba/ilogtail/assets/101638329/ef4b4007-4796-4e43-aca2-452aa70fa016">
<img width="1450" alt="image" src="https://github.com/alibaba/ilogtail/assets/101638329/7f11e851-91a1-452d-8629-80bd1354db9c">
从日志中依然未发现任何信息

ilogtail_config.json配置：
<img width="726" alt="image" src="https://github.com/alibaba/ilogtail/assets/101638329/9d00e7f7-a6ba-4259-98f2-74d2cd92ab43">

**Reply by linrunqi08** [reply]
@q121102302  这个case中的原始日志文件能否提供一下呢，我们用这个文件复现一下

**Reply by q121102302** [reply]
其实跟上面发的那个文件没什么区别的。。。公司内外网分离 拿原始日志很麻烦

**Reply by shalousun** [reply]
@q121102302 如果可以准确复现可以在flusher 上再加个stdout输出辅助排查下，看下数据是在哪一层丢的，stdout打印了丢失日志，那可能flusher kafka层丢了数据。先将范围确定下

**Reply by linrunqi08** [reply]
https://github.com/alibaba/ilogtail/discussions/828  有点像这个问题，如果是单行日志的话，可以把processor_split_log_regex这个插件去掉看下

**Reply by q121102302** [reply]
今天早上修改配置 在user_yaml_config.d下的配置文件中新增
`- Type: flusher_stdout
    OnlyStdout: false
    MaxSize: 2000000000
    FileName: /opt/ilogtail/test.log`

下午依然出现大日志未成功的问题  这个日志输出在早上的重启ilogtail后 测试三次都是正常的，下午测试就失败了

ilogtail日志：
ilogtail.LOG
<img width="1610" alt="image" src="https://github.com/alibaba/ilogtail/assets/101638329/addc7019-6897-4968-9f9d-a44a750a6801">
<img width="1641" alt="image" src="https://github.com/alibaba/ilogtail/assets/101638329/ee79e854-ea7b-448d-be24-ba6d439cefbb">

logtail_plugin.LOG
<img width="1653" alt="image" src="https://github.com/alibaba/ilogtail/assets/101638329/37f4cb7c-85a0-4bfb-ab1c-c3674250da8d">


原始日志：红框内为丢失的数据 箭头指向的为那条大日志
<img width="1661" alt="image" src="https://github.com/alibaba/ilogtail/assets/101638329/09b9d200-d7d1-48e8-b260-6494b0936927">

stdout输出的日志：
<img width="1597" alt="image" src="https://github.com/alibaba/ilogtail/assets/101638329/ce932b4f-ec55-4cf9-bbbf-9208d3e29619">
可以看到直接从2023-05-31 13:45:49.144直接跳到了2023-05-31 13:45:55.434

**Reply by q121102302** [reply]
processor_split_log_regex 去掉这个插件 本身也会让我们错误日志不方便查询呀

**Comment by linrunqi08** [comment, marked as answer]
duplicate to https://github.com/alibaba/ilogtail/discussions/828, will trace this problem in https://github.com/alibaba/ilogtail/discussions/828
